File: OS/Python/08_best_fit.py
# Memory Allocation — Best Fit

nb = int(input("Number of blocks: "))
bs = [int(input(f"Block {i+1} size: ")) for i in range(nb)]
np = int(input("Number of processes: "))
ps = [int(input(f"Process {i+1} size: ")) for i in range(np)]

alloc = [-1] * np
used  = [False] * nb

for j in range(np):
    best, best_idx = float('inf'), -1
    for i in range(nb):
        diff = bs[i] - ps[j]
        if not used[i] and 0 <= diff < best:
            best, best_idx = diff, i
    if best_idx != -1:
        alloc[j] = best_idx; used[best_idx] = True

print(f"\n{'Process':<10}{'P.Size':<10}{'Block':<8}{'B.Size':<10}{'Fragment'}")
print("-" * 46)
for j in range(np):
    if alloc[j] != -1:
        frag = bs[alloc[j]] - ps[j]
        print(f"P{j+1:<9}{ps[j]:<10}{alloc[j]+1:<8}{bs[alloc[j]]:<10}{frag}")
    else:
        print(f"P{j+1:<9}{ps[j]:<10}Not Allocated")

File: OS/C/08_Best_Fit.c
/*
 * Program  : Memory Allocation — Best Fit
 * Subject  : Operating Systems Lab
 *
 * Description:
 *   In the Best Fit strategy, each process is allocated to the SMALLEST
 *   block that is still large enough to accommodate it.
 *
 *   Advantage  : Minimizes wasted space (internal fragmentation) per allocation.
 *   Disadvantage: Slower — must scan all blocks. Leaves many tiny unusable gaps.
 */

#include <stdio.h>

#define MAX_SIZE 25

int main() {
    int num_blocks, num_processes;
    int block_size[MAX_SIZE];
    int process_size[MAX_SIZE];
    int fragment[MAX_SIZE];     /* Fragmentation for each process */
    int block_used[MAX_SIZE];   /* 1 = block already allocated, 0 = free */
    int fit_block[MAX_SIZE];    /* Which block each process was assigned to */

    int i, j;
    int temp_diff;
    int smallest_gap;           /* Tracks the minimum leftover for best fit */

    /* Initialize arrays */
    static int zero_init[MAX_SIZE];
    for (i = 0; i < MAX_SIZE; i++) {
        block_used[i] = 0;
        fit_block[i]  = 0;
        fragment[i]   = 0;
    }

    /* --- Input --- */
    printf("\nEnter the number of blocks: ");
    scanf("%d", &num_blocks);
    printf("Enter the number of processes: ");
    scanf("%d", &num_processes);

    printf("\nEnter the size of each block:\n");
    for (i = 1; i <= num_blocks; i++) {
        printf("  Block %d: ", i);
        scanf("%d", &block_size[i]);
    }

    printf("Enter the size of each process:\n");
    for (i = 1; i <= num_processes; i++) {
        printf("  Process %d: ", i);
        scanf("%d", &process_size[i]);
    }

    /* --- Best Fit Allocation --- */
    for (i = 1; i <= num_processes; i++) {
        smallest_gap = 10000;  /* Large initial value */

        for (j = 1; j <= num_blocks; j++) {
            if (block_used[j] == 1)
                continue;  /* Block already taken */

            temp_diff = block_size[j] - process_size[i];

            /* Valid fit: block is large enough AND is the best (smallest gap) so far */
            if (temp_diff >= 0 && temp_diff < smallest_gap) {
                fit_block[i]  = j;
                smallest_gap  = temp_diff;
            }
        }

        if (fit_block[i] != 0) {
            fragment[i]          = smallest_gap;
            block_used[fit_block[i]] = 1;  /* Mark block as used */
        }
    }

    /* --- Output Table --- */
    printf("\n%-12s %-14s %-14s %-12s %-12s\n",
           "Process", "Process Size", "Block No", "Block Size", "Fragment");
    printf("-----------------------------------------------------------------\n");

    for (i = 1; i <= num_processes; i++) {
        if (fit_block[i] != 0) {
            printf("%-12d %-14d %-14d %-12d %-12d\n",
                   i, process_size[i], fit_block[i],
                   block_size[fit_block[i]], fragment[i]);
        } else {
            printf("%-12d %-14d Not Allocated\n", i, process_size[i]);
        }
    }

    return 0;
}
